File: Java/Migration/unit-tests-csharp/unit-tests-csharp/P04NSubstitute/Ex03ThrowingExceptionsFromMocks.cs
using System;
using NSubstitute;
using NSubstitute.ExceptionExtensions;
using NUnit.Framework;
using TddEbook.TddToolkit;
using unit_tests_csharp.P04NSubstitute.ProductionCode;

public class Ex03ThrowingExceptionsFromMocks
{

  [Test]
  public void ShouldThrowExceptionWhenReadingFromSourceThrowsException()
  {
    //GIVEN
    var copyOperation = new CopyOperation();
    var destination = Substitute.For<IDataDestination>();
    var source = Substitute.For<IDataSource>();
    var exception = Any.Exception();

    source.RetrieveData().Throws(exception);

    //WHEN - THEN
    var e = Assert.Throws<Exception>(
      () => copyOperation.ApplyTo(source, destination));
    Assert.AreEqual(exception, e);
    //never verification
    destination.DidNotReceive().Save(Arg.Any<Data>());
  }

  [Test]
  public void ShouldThrowExceptionWhenSavingToDestinationThrowsException()
  {
    //GIVEN
    var copyOperation = new CopyOperation();
    var destination = Substitute.For<IDataDestination>();
    var source = Substitute.For<IDataSource>();
    var data = Any.Instance<Data>();
    var exception = Any.Exception();

    source.RetrieveData().Returns(data);
    destination.When(d => d.Save(data)).Throw(exception);

    //WHEN - THEN
    var thrownException = Assert.Throws<Exception>(
      () => copyOperation.ApplyTo(source, destination));
    Assert.AreEqual(exception, thrownException);
  }


}

File: Java/Migration/unit-tests-java/src/test/java/p04_mocks/Ex03ThrowingExceptionsFromMocks.java
package p04_mocks;

import autofixture.publicinterface.Any;
import org.testng.annotations.Test;
import p04_mockito.CopyOperation;
import p04_mockito.Data;
import p04_mockito.DataDestination;
import p04_mockito.DataSource;

import static org.assertj.core.api.AssertionsForClassTypes.assertThatThrownBy;
import static org.mockito.Mockito.*;

public class Ex03ThrowingExceptionsFromMocks {

  @Test
  public void shouldThrowExceptionWhenReadingFromSourceThrowsException() {
    //GIVEN
    CopyOperation copyOperation = new CopyOperation();
    DataDestination destination = mock(DataDestination.class);
    DataSource source = mock(DataSource.class);
    RuntimeException exception = Any.anonymous(RuntimeException.class);

    when(source.retrieveData()).thenThrow(exception);

    //WHEN - THEN
    assertThatThrownBy(() -> copyOperation.applyTo(source, destination))
        .isEqualTo(exception);
    //never verification
    verify(destination, never()).save(any(Data.class));
  }

  @Test
  public void shouldThrowExceptionWhenSavingToDestinationThrowsException() {
    //GIVEN
    CopyOperation copyOperation = new CopyOperation();
    DataDestination destination = mock(DataDestination.class);
    DataSource source = mock(DataSource.class);
    Data data = Any.anonymous(Data.class);
    RuntimeException exception = Any.anonymous(RuntimeException.class);

    when(source.retrieveData()).thenReturn(data);
    doThrow(exception).when(destination).save(data);

    //WHEN - THEN
    assertThatThrownBy(() -> copyOperation.applyTo(source, destination))
        .isEqualTo(exception);
  }


}
